internal/charmstore/store.go

```diff
@@ -505,6 +505,84 @@ func (s *Store) UpdateBaseEntity(url *charm.Reference, update interface{}) error
 	return nil
 }
 
+// Promulgate sets the base entity of url to be promulgated, and unsets promulgated
+// on any other base entity for charms with the same name. It also calculates the next
+// promulgated URL for the charms owned by the new owner and sets those charms.
+//
+// Note: This code is known to have some unfortunate (but not dangerous) race
+// conditions, it is considered that the promulgate operation will be significantly rare
+// that it is unlikely thay will occur.
+func (s *Store) Promulgate(url *charm.Reference) error {
+	entity, err := s.FindEntity(url)
+	if err != nil {
+		return errgo.Mask(err, errgo.Any)
+	}
+	bURL := baseURL(entity.URL)
+	_, err = s.DB.BaseEntities().UpdateAll(
+		bson.D{{"$or", []bson.D{
+			{{"$and", []bson.D{{{"_id", bURL}}, {{"promulgated", -1}}}}},
+			{{"$and", []bson.D{{{"name", bURL.Name}}, {{"_id", bson.D{{"$ne", bURL}}}}, {{"promulgated", 1}}}}},
+		}}},
+		bson.D{{"$mul", bson.D{{"promulgated", -1}}}},
+	)
+	if err != nil {
+		return errgo.Mask(err)
+	}
+	type result struct {
+		Series   string `bson:"_id"`
+		Revision int
+	}
+	latestOwned := make(map[string]int)
+	iter := s.DB.Entities().Pipe([]bson.D{
+		{{"$match", bson.D{{"baseurl", bURL}}}},
+		{{"$group", bson.D{{"_id", "$series"}, {"revision", bson.D{{"$max", "$revision"}}}}}},
+	}).Iter()
+	var res result
+	for iter.Next(&res) {
+		latestOwned[res.Series] = res.Revision
+	}
+	if err := iter.Close(); err != nil {
+		return errgo.Mask(err)
+	}
+	latestPromulgated := make(map[string]int)
+	iter = s.DB.Entities().Pipe([]bson.D{
```

Comment:
// Find the latest promulgated URL revision for all entities that share
    // the same name. Note that this is works because:
    // 1) promulgated URLs always have the same charm name as their non-promulgated
    //    counterparts.
    // 2) bundles cannot have names that overlap with charms.
    //  Because of 1), we are sure that selecting on the entity name will
    // select all entities with a matching promulgated URL name.
    // Because of 2) we are sure that we are only updating all charms or the
    // single bundle entity.

?